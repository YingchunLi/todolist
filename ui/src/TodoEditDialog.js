import React, {Component} from 'react';

import _ from 'lodash';

import DropDownMenu from 'material-ui/DropDownMenu';
import MenuItem from 'material-ui/MenuItem';

import TextField from 'material-ui/TextField';
import {
  Table,
  TableBody,
  TableRow,
  TableRowColumn,
} from 'material-ui/Table';

import Dialog from 'material-ui/Dialog';
import FlatButton from 'material-ui/FlatButton';
import ActionEdit from 'material-ui/svg-icons/editor/mode-edit';
import ActionAdd from 'material-ui/svg-icons/content/add-circle';

import DateTimePicker from 'material-ui-datetimepicker';
import DatePickerDialog from 'material-ui/DatePicker/DatePickerDialog'
import TimePickerDialog from 'material-ui/TimePicker/TimePickerDialog';

class TodoEditDialog extends Component {
  state = {
    open: false,
    todo: this.props.todo && _.cloneDeep(this.props.todo),
  };

  componentWillReceiveProps(nextProps, nextContext) {
    if (!_.isEqual(this.props.todo, nextProps.todo)) {
      this.setState({todo: _.cloneDeep(nextProps.todo)});
    }
  }

  handleOpen = () => {
    this.setState({open: true});
  };

  handleClose = () => {
    this.setState({open: false, todo: undefined});
  };

  handleSubmit = () => {
    this.props.onChange(this.state.todo);
    this.handleClose();
  };

  handleChange = (field, value) => {
    const newTodo = {...this.state.todo, [field]: value};
    this.setState({todo: newTodo});
  };


  render() {
    const {mode, disabled} = this.props;
    const {todo={}} = this.state;
    const actions = [
      <FlatButton
        label="Cancel"
        primary={true}
        onClick={this.handleClose}
      />,
      <FlatButton
        label="Submit"
        primary={true}
        keyboardFocused={true}
        onClick={this.handleSubmit}
      />,
    ];

    const todoEditTable =
        <Table >
          <TableBody displayRowCheckbox={false} showRowHover={true}>
            <TableRow>
              <TableRowColumn style={{width: '30%'}}>Name</TableRowColumn>
              <TableRowColumn>
                <TextField
                  id="name"
                  value={todo.name || ''}
                  onChange={(event) => this.handleChange('name', event.target.value)}
                />
                </TableRowColumn>
            </TableRow>
            <TableRow>
              <TableRowColumn>Description</TableRowColumn>
              <TableRowColumn>
                <TextField
                  id="description"
                  value={todo.description || ''}
                  multiLine={true}
                  rows={3}
                  onChange={(event) => this.handleChange('description', event.target.value)}
                />
              </TableRowColumn>
            </TableRow>
            <TableRow>
              <TableRowColumn>Due Date</TableRowColumn>
              <TableRowColumn>
                <DateTimePicker
                  value={todo.dueDate ? new Date(todo.dueDate) : undefined}
                  onChange={(dateTime) => this.handleChange('dueDate', dateTime.toISOString())}
                  DatePicker={DatePickerDialog}
                  TimePicker={TimePickerDialog}
                />
              </TableRowColumn>
            </TableRow>
            <TableRow>
              <TableRowColumn>Status</TableRowColumn>
              <TableRowColumn>
                <DropDownMenu maxHeight={300}
                              value={todo.status || 'Pending'}
                              onChange={(event, index, value) => this.handleChange('status', value)}
                              disabled={mode === 'Add'}
                >
                  <MenuItem value="Pending" key="Pending" primaryText="Pending" />
                  <MenuItem value="Done" key="Done" primaryText="Done" />
                </DropDownMenu>
              </TableRowColumn>
            </TableRow>
          </TableBody>
        </Table>;

    return (
      <div>
        <FlatButton
          label={mode}
          icon={mode === 'Add' ? <ActionAdd /> : <ActionEdit />}
          onClick={this.handleOpen}
          disabled={disabled}
        />

        <Dialog
          title={`${mode} Todo`}
          actions={actions}
          modal={false}
          open={this.state.open}
          onRequestClose={this.handleClose}
          autoScrollBodyContent={true}
        >
          {todoEditTable}
        </Dialog>
      </div>
    );
  }
}

TodoEditDialog.propTypes = {};
TodoEditDialog.defaultProps = {};

export default TodoEditDialog;
