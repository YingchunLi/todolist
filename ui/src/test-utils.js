import React from 'react';
import { MemoryRouter as Router } from 'react-router-dom';

import configureStore from './store';

import {Provider} from 'react-redux';

// wrapper for store
export const withStore = (WrappedComponent, state = {}) => (props) => {
  const store = configureStore(state);
  return (
    <Router>
      <Provider store={store}>
        <WrappedComponent {...props} />
      </Provider>
    </Router>
  );
};

