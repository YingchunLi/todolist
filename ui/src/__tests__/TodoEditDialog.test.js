import React from "react";
import {MemoryRouter, Route} from "react-router-dom";

import {shallow, mount} from "enzyme";
import renderer from "react-test-renderer";

import TodoEditDialog from "../TodoEditDialog";
import {withStoreAndMaterialUI} from '../test-utils';

const TodoEditDialogWithStoreAndMaterialUI = withStoreAndMaterialUI(TodoEditDialog);

describe("Todo", () => {


    it("not todo - add mode", () => {
      const component = renderer.create(
        <TodoEditDialogWithStoreAndMaterialUI mode="Add" />
      ).toJSON();
      expect(component).toMatchSnapshot();
    });

    it("has todo - edit mode", () => {
      const component = renderer.create(
        <TodoEditDialogWithStoreAndMaterialUI
          mode="Edit"
          todo={
            {
              id: 1,
              name: 'milk',
              description: 'remember the milk',
              dueDate: '2018-12-31',
              status: 'Pending'
            }
          }
          onChange={() => {}}
        />
      ).toJSON();
      expect(component).toMatchSnapshot();
    });

  }
);

