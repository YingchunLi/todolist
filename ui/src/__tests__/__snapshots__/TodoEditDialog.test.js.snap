// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo has todo - edit mode 1`] = `
<div>
  <button
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
        "backgroundColor": "rgba(0, 0, 0, 0)",
        "border": 10,
        "borderRadius": 2,
        "boxSizing": "border-box",
        "color": "rgba(0, 0, 0, 0.87)",
        "cursor": "pointer",
        "display": "inline-block",
        "fontFamily": "Roboto, sans-serif",
        "fontSize": "inherit",
        "fontWeight": "inherit",
        "height": 36,
        "lineHeight": "36px",
        "margin": 0,
        "minWidth": 88,
        "muiPrepared": true,
        "outline": "none",
        "overflow": "hidden",
        "padding": 0,
        "position": "relative",
        "textAlign": "center",
        "textDecoration": "none",
        "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
        "userSelect": "none",
        "verticalAlign": null,
      }
    }
    tabIndex={0}
    type="button"
  >
    <div
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
    >
      <svg
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "display": "inline-block",
            "fill": "rgba(0, 0, 0, 0.87)",
            "height": 24,
            "marginLeft": 12,
            "marginRight": 0,
            "muiPrepared": true,
            "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
            "userSelect": "none",
            "verticalAlign": "middle",
            "width": 24,
          }
        }
        viewBox="0 0 24 24"
      >
        <path
          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.39-.39-1.02-.39-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
        />
      </svg>
      <span
        style={
          Object {
            "fontSize": 14,
            "fontWeight": 500,
            "letterSpacing": 0,
            "muiPrepared": true,
            "paddingLeft": 8,
            "paddingRight": 16,
            "position": "relative",
            "textTransform": "uppercase",
            "verticalAlign": "middle",
          }
        }
      >
        Edit
      </span>
    </div>
  </button>
</div>
`;

exports[`Todo not todo - add mode 1`] = `
<div>
  <button
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
        "backgroundColor": "rgba(0, 0, 0, 0)",
        "border": 10,
        "borderRadius": 2,
        "boxSizing": "border-box",
        "color": "rgba(0, 0, 0, 0.87)",
        "cursor": "pointer",
        "display": "inline-block",
        "fontFamily": "Roboto, sans-serif",
        "fontSize": "inherit",
        "fontWeight": "inherit",
        "height": 36,
        "lineHeight": "36px",
        "margin": 0,
        "minWidth": 88,
        "muiPrepared": true,
        "outline": "none",
        "overflow": "hidden",
        "padding": 0,
        "position": "relative",
        "textAlign": "center",
        "textDecoration": "none",
        "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
        "userSelect": "none",
        "verticalAlign": null,
      }
    }
    tabIndex={0}
    type="button"
  >
    <div
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
    >
      <svg
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "display": "inline-block",
            "fill": "rgba(0, 0, 0, 0.87)",
            "height": 24,
            "marginLeft": 12,
            "marginRight": 0,
            "muiPrepared": true,
            "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
            "userSelect": "none",
            "verticalAlign": "middle",
            "width": 24,
          }
        }
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
        />
      </svg>
      <span
        style={
          Object {
            "fontSize": 14,
            "fontWeight": 500,
            "letterSpacing": 0,
            "muiPrepared": true,
            "paddingLeft": 8,
            "paddingRight": 16,
            "position": "relative",
            "textTransform": "uppercase",
            "verticalAlign": "middle",
          }
        }
      >
        Add
      </span>
    </div>
  </button>
</div>
`;
